### PR TITLE
Update error handling as per Go 1.13 guidelines

### PR DESCRIPTION
**Pull request by mbalayil**
Fixes [1846](https://github.com/google/go-github/issues/1846)

1. Updated `README.md` to reflect that it is no longer Go 1.9 that we require, but instead we need a minimum of Go 1.13.
2. Replaced `reflect.DeepEqual` with `errors.Is` in all the affected tests.
3. Implemented `Is(error) bool` interface for all the required errors (`ErrorResponse`, `RateLimitError`, `AcceptedError` and `AbuseRateLimitError`). All the individual fields are directly compared. However, for the `*http Response` field, only `StatusCode` is being compared. If the other fields of the `http.Response` struct need to be compared, please let me know.

### CODE REVIEW

File: github/github.go
Comment (gmlewis) [inline github/github.go:689]: I'm not sure if you can see all the warnings I see on GitHub from Codecov, but please add a unit test that covers this line.  I'll highlight the others in case you can't see them.

File: github/github.go
Comment (gmlewis) [inline github/github.go:692]: Please add a unit test that covers this line.

File: github/github.go
Comment (gmlewis) [inline github/github.go:686]: Please add a comment for these new methods... something like:

```
// Is tests if the target error is of type ErrorResponse.
```

File: github/github.go
Comment (gmlewis) [inline github/github.go:696]: Please add a unit test that covers this line.

File: github/github.go
Comment (gmlewis) [inline github/github.go:700]: Please add a unit test that covers this line.

File: github/github.go
Comment (gmlewis) [inline github/github.go:704]: Please add a unit test that covers this line.

File: github/github.go
Comment (gmlewis) [inline github/github.go:710]: Please add a unit test that covers this line.

File: github/github.go
Comment (gmlewis) [inline github/github.go:737]: Same here, please.

File: github/github.go
Comment (gmlewis) [inline github/github.go:740]: Please add a unit test that covers this line.

File: github/github.go
Comment (gmlewis) [inline github/github.go:743]: Please add a unit test that covers this line.

File: github/github.go
Comment (gmlewis) [inline github/github.go:747]: Please add a unit test that covers this line.

File: github/github.go
Comment (gmlewis) [inline github/github.go:768]: Same here, please.

File: github/github.go
Comment (gmlewis) [inline github/github.go:771]: Please add a unit test that covers this line.

File: github/github.go
Comment (gmlewis) [inline github/github.go:773]: Please add a unit test that covers this line.

File: github/github.go
Comment (gmlewis) [inline github/github.go:794]: Same here, please.

File: github/github.go
Comment (gmlewis) [inline github/github.go:797]: Please add a unit test that covers this line.

File: github/github.go
Comment (gmlewis) [inline github/github.go:800]: Please add a unit test that covers this line.

File: github/github.go
Comment (gmlewis) [inline github/github.go:804]: Please add a unit test that covers this line.

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @mbalayil !
Please address the comments below.

**Comment by mbalayil** [comment]
Have updated the PR with more tests to improve coverage. Please take a look.

File: github/github.go
Comment (mbalayil) [reply github/github.go:701]: Done.

File: github/github.go
Comment (mbalayil) [reply github/github.go:704]: I have addressed all code coverage issues with additional tests.

File: github/github.go
Comment (mbalayil) [reply github/github.go:709]: Done.

File: github/github.go
Comment (mbalayil) [reply github/github.go:718]: Done.

File: github/github.go
Comment (mbalayil) [reply github/github.go:732]: Done.

File: github/github.go
Comment (mbalayil) [reply github/github.go:766]: Done.

File: github/github.go
Comment (mbalayil) [reply github/github.go:769]: Done.

File: github/github.go
Comment (mbalayil) [reply github/github.go:793]: Done.

File: github/github.go
Comment (mbalayil) [reply github/github.go:796]: Done.

File: github/github.go
Comment (mbalayil) [reply github/github.go:798]: Done.

File: github/github.go
Comment (mbalayil) [reply github/github.go:820]: Done.

File: github/github.go
Comment (mbalayil) [reply github/github.go:823]: Done.

File: github/github.go
Comment (mbalayil) [reply github/github.go:714]: Done.

**Review by gmlewis — APPROVED** [review]
Thank you, @mbalayil !
LGTM.

Awaiting second LGTM before merging.

**Review by wesleimp — APPROVED** [review]
💙

**Comment by gmlewis** [comment]
Thank you, @wesleimp!

@willnorris - are you OK with this implementation?  I'll wait until we hear from you... no rush... I just want to make sure since you had some comments on the thread.

**Comment by willnorris** [comment]
> @willnorris - are you OK with this implementation?

My biggest concern, just looking at this from a high level, is that this is a LOT of work and extra code just to be able to call `errors.Is` in a couple of test cases.  If having these errors be comparable in this way has real value outside of just these tests, then that may be fine, and it may be worth the effort.  But if it's **just** to have marginally cleaner tests (not calling `reflect.DeepEqual`), I find it harder to justify.

The whole point of the new go1.13 error handling is that the `Is` and `As` functions support wrapped errors.  Both of them will intelligently unwrap errors to try and find a match.  But we're not wrapping errors.  Our errors are *just values*, like any other value in Go.  So honestly, there's no advantage of using `errors.Is` versus the go-cmp package (in our tests, I mean).

I really appreciate all the effort @mbalayil put into this... really, I do.  But looking at what it ended up taking in the end, I'm not sure that this is the right solution for this problem.  (Unless there's some other use case that this is satisfying, outside of those two test cases.)

**Comment by gmlewis** [comment]
Correct me if I'm wrong, @willnorris , but won't the availability of the `Is` methods (added by tihs PR) make it easier for the users of this package to wrap _their_ errors and still test if the base error is one of the types defined in this repo?

In other words, doesn't the work here make error handler easier for future users of this package (and also make the usage more inline with the [Go 1.13 Error Handling blog post](https://blog.golang.org/go1.13-errors))?

**Comment by willnorris** [comment]
Yeah, that's a good point.  For the sentinel errors we have, I think that will be very useful (but also doesn't require a custom `Is` method).  Given how a couple of these errors are tied to very specific http requests/responses, I think the full equality check is going to be of very limited value, even for downstream users, but am happy to be proven wrong.

**Review by willnorris — APPROVED** [review]

**Comment by gmlewis** [comment]
> I think the full equality check is going to be of very limited value, even for downstream users, but am happy to be proven wrong.

So do you think it would be more useful _without_ the equality checks?

Or should we run with this for awhile until we get feedback and then adjust accordingly?

**Comment by willnorris** [comment]
Looking more at the examples I was concerned about, they seem to be more limited than I thought.  I was mostly focusing on the `CreatedAt` timestamp that exists in some errors.  I thought that was more common, but I now see it is only for HTTP 451 errors.  For those errors, an equality check will almost never match in the real world because you won't know the creation date ahead of time.

I guess I'd need to look at some more real-world errors to see how likely it would be for people to really be doing equality checks.  My guess is that most go-github users that want to wrap our errors and eventually test for error types will use `errors.As` (which we don't have to do anything for).

I know that doesn't directly answer your question, but I guess I'm discovering that my once somewhat-strongly-held view is softening considerably, so I can go either way on this now.

**Comment by gmlewis** [comment]
OK, sounds good.  Thanks, Will! 
Let's run with this and see what kind of feedback we get.  :joy:
Merging.

**Pull request merged**
